### ISSUE TITLE
Questions for hw5

### ISSUE THREAD

**Issue by littlehug** [SytemProgramming2015FallatNTU/Announcement#81]
Hi, I am not sure for some requisitions in detail, so here are some questions to ask.

> The pidfile will contain the pid of current running csiebox_server. The recorded pid will be used by start-stop script to stop the process.

For Daemonize part, we are asked to edit the content of pidfile, Does the program **append** the pid to file or just **override** it?

> After catching the signal, server will write how many threads in the thread pool are active to the fifo between server process and web process. Notice: Before writing an integer into the file, you should pass the integer into `htonl`. This is for endianness. Moreover, you should write 4 bytes by using **uint32_t** instead of int.

For Active Threads Statics part,  
- Not quite to know definition of *active thread*. Does the thread that is active to the fifo between server process and web process means it is a **free** thread? or it is a thread which is processing for fifo.pid? or other meaning?
- After `htonl`, do we need to write uint32_t directly by binary, or convert uint32_t to string?

Thanks a lot !

**Comment by kucw** [comment]
1. pidfile will be created by csiebox_server.sh
2. active thread means **busy** thread in MP4, for example, if you receive a upload request, you have to choose a thread in the thread pool to handle this request, and this thread is called "busy" thread in MP4, also  called  "active" thread here.
3. use `htonl` to convert **uint32_t** from host byte order to network byte order, and write it directly to fifo.

**Comment by littlehug** [comment]
>However, during the process of daemonization, the pid will be changed. You should update the pidfile to make the start-stop script can stop the csiebox_server correctly.

Don't we need to change the content of pidfile ?
Thanks.

**Comment by kucw** [comment]
Yes, the script will kill the process by the pid which is in the pidfile, so you need to change the content of pidfile because the pid of csiebox_server will be changed after daemonization.

**Comment by littlehug** [comment]
May I ask a question about "MP4 answer" released recently?

``` c
typedef struct {
  struct {
    char path[PATH_MAX];
    char account_path[PATH_MAX];
    int thread;
    char run_path[PATH_MAX];
    int daemonize;
  } arg;
  char fifo[PATH_MAX];
  int listen_fd;
  csiebox_client_info** client;
  thread_pool* threads;
} csiebox_server;
```
For the file "csiebox_server.h" in include/ directory, there is an additional string stored in csiebox_server structure named **fifo**. I guess it is to store fifo.[pid] 's filename.
I am wondering when we will use it?
It seems to be unused for fifo while we are going to write the number of active thread after we catch SIGUSR1,  since signal_handler can accept only one integer as argument. There is no opportunity for us to pass csiebox_server type argument. OR this implies that we don't need to use signal_handler to deal with the file?
Is the pid filled in fifo.[pid] filename the parent pid or child pid after fork()?
Thanks!

**Comment by kucw** [comment]
You can use `static csiebox_server* server;` to store fifo path and other information togother, or use `static char fifo[PATH_MAX];` to store fifo path, is all up to you.
Fifo.[pid] is the way for server to communicate with web system, and the parent will be killed after daemonize, so the pid should filled with child's pid. You can reference the suggested structure of MP5 [here](https://github.com/SytemProgramming2015FallatNTU/Announcement/wiki/Homework-5#second-part-daemonize).

**Comment by littlehug** [comment]
Thank you.
Forget to ask, do we need to process "server busy" and file lock in MP5?
I cannot find any output about these function in released answer.

**Comment by kucw** [comment]
The released answer is the simple version of MP4, so it don't contain server busy and file lock. In MP5, we want you practice IPC, so the test case won't have server busy and file lock situation, and you can just print the number of active thread in MP4 answer to web system.
